File: egress_hub/code/me_egress_poller/src/main/java/digital/exchange/egress/hub/me/egress/poller/MainLoopAgent.java
package digital.exchange.egress.hub.me.egress.poller;

import digital.exchange.egress.hub.me.egress.poller.hub.LocalResponseChannelPublication;
import digital.exchange.egress.hub.me.egress.poller.subscription.fsm.SubscriptionStateAgent;
import digital.exchange.sbe.hub.inner.SignalEnumFromHubNodeDecoder;
import digital.exchange.sbe.hub.inner.MeMessageEncoder;
import digital.exchange.sbe.hub.inner.MessageHeaderEncoder;
import io.aeron.logbuffer.Header;
import digital.exchange.egress.hub.me.egress.poller.hub.SignalDecoder;
import digital.exchange.egress.hub.me.egress.poller.me.egress.archive.MeEgressArchiveClientStateAgent;
import digital.exchange.egress.hub.me.egress.poller.cluster.client.fsm.ClusterClientStateAgent;
import digital.exchange.egress.hub.me.egress.poller.me.egress.MeEgressListener;
import digital.exchange.egress.hub.me.egress.poller.utils.AppConfig;
import org.agrona.DirectBuffer;
import org.agrona.ExpandableDirectByteBuffer;
import org.agrona.concurrent.Agent;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.util.List;

public class MainLoopAgent implements Agent
{
    private static final Logger LOGGER = LoggerFactory.getLogger(MainLoopAgent.class);
    private final SubscriptionStateAgent hubSubscriptionStateAgent = new SubscriptionStateAgent(
            AppConfig.hubInnerSignalChannel,
            AppConfig.hubInnerSignalChannelStreamId,
            this::hubSubscriptionOnFragment
    );

    private final LocalResponseChannelPublication publication = new LocalResponseChannelPublication();

    private final MeEgressArchiveClientStateAgent meEgressArchiveClientStateAgent =
            new MeEgressArchiveClientStateAgent(this::meEgressArchiveOnFragment);

    private final MeEgressListener meEgressListener = new MeEgressListener(
            meEgressArchiveClientStateAgent::onMeNewLeader);
    private final ClusterClientStateAgent meClusterClientStateAgent = new ClusterClientStateAgent(
            List.of(AppConfig.meHosts.split(",")),
            AppConfig.meClusterPortBase,
            AppConfig.meClusterClientEgressChannel,
            AppConfig.meClusterIngressChannel,
            meEgressListener,
            meEgressArchiveClientStateAgent::onMeNewLeader
    );

    private final ExpandableDirectByteBuffer buffer = new ExpandableDirectByteBuffer();
    private long bufferMsgPosition = -1;
    private boolean bufferUpdatedFlag = false;
    private boolean isMsgSent = false;
    private final MessageHeaderEncoder messageHeaderEncoder = new MessageHeaderEncoder();
    private final MeMessageEncoder meMessageEncoder = new MeMessageEncoder();

    private void hubSubscriptionOnFragment(DirectBuffer buffer, int offset, int length, Header header)
    {
        var signal = SignalDecoder.decodeSignal(buffer, offset, length);

        if(signal != null)
        {
            dispatchSignal(signal);
        }
    }

    private void dispatchSignal(final SignalEnumFromHubNodeDecoder signalDecoder)
    {
        final long position = signalDecoder.position();

        switch (signalDecoder.signal())
        {
            case CONNECT -> {
                meClusterClientStateAgent.connectToCluster();
                meEgressArchiveClientStateAgent.connectToMeEgressArchive(position);
            }

            case DISCONNECT -> {
                meClusterClientStateAgent.disconnectFromCluster();
                meEgressArchiveClientStateAgent.disconnectFromMeEgressArchive();
            }

            case DO_WORK -> {
                if(bufferMsgPosition != position)
                {
                    bufferMsgPosition = position;
                    bufferUpdatedFlag = false;
                    isMsgSent = false;
                    meEgressArchiveClientStateAgent.doWork(position);
                }
                else if(!bufferUpdatedFlag)
                {
                    meEgressArchiveClientStateAgent.doWork(position);
                }
                else if(!isMsgSent)
                {
                    sendBuffer();
                }
            }
        }

    }

    private void meEgressArchiveOnFragment(final DirectBuffer buffer, final int offset, final int length, final Header header)
    {
        LOGGER.info("Offering Msg to Cluster, offset: {}, length: {}, position: {}", offset, length, header.position());

        meMessageEncoder.wrapAndApplyHeader(this.buffer, 0, messageHeaderEncoder)
                .position(header.position())
                .putMessageBytes(buffer, offset, length);

        bufferUpdatedFlag = true;

        sendBuffer();
    }

    private void sendBuffer()
    {
        var res = publication.offer(this.buffer, 0,
                MessageHeaderEncoder.ENCODED_LENGTH + meMessageEncoder.encodedLength());

        if(res > 0)
        {
            isMsgSent = true;
        }
    }

    @Override
    public void onStart() {
        LOGGER.info("onStart");
    }

    @Override
    public int doWork() {
        int ret = hubSubscriptionStateAgent.doWork();
        ret += meClusterClientStateAgent.doWork();

        return ret;
    }

    @Override
    public void onClose() {
        LOGGER.info("onClose");
    }

    @Override
    public String roleName() {
        return "MeEgressArchiveClientStateAgent";
    }

}

File: egress_hub/code/cache_order_book_updater/MainLoopAgent/MainLoopAgent.cpp
#include "MainLoopAgent.h"

#include <spdlog/sinks/stdout_color_sinks.h>

#include "AppConfig.h"

MainLoopAgent::MainLoopAgent():
signalSubsciption(
    [this](
        aeron::concurrent::AtomicBuffer &buffer,
        aeron::util::index_t offset,
        aeron::util::index_t length,
        aeron::Header &header
    )
    {
        this->fragmentHandler(buffer, offset, length, header);
    }
)
{
    logger = spdlog::stdout_color_mt("MainLoopAgent");
    logger->set_pattern(AppConfig::loggerPatern);
}

void MainLoopAgent::fragmentHandler
(
    aeron::concurrent::AtomicBuffer &buffer,
    aeron::util::index_t offset,
    aeron::util::index_t length,
    aeron::Header &header
)
{
    auto signal_ptr = signalDecoder.decodeSignal(reinterpret_cast<char*>(buffer.buffer()), offset, length);
    if(signal_ptr != nullptr)
    {
        dispatchSignal(signal_ptr->signal, signal_ptr->position);
    }
}

void MainLoopAgent::dispatchSignal(const SignalCommandEnum &signal, const int64_t position)
{
    switch (signal)
    {
    case CONNECT:
        logger->info("CONNECT");
        break;
    
    case DISCONNECT:
        logger->info("DISCONNECT");
        break;

    case DO_WORK:
        if(position > lastPosition)
        {
            if(readMessageAndPushToCache())
            {
                lastPosition = position;
                responsePublication.sendPositionUpdate(position);
            }
        }
        break;
    
    default:
        break;
    }
}

bool MainLoopAgent::readMessageAndPushToCache()
{
    if(mappedBuffer_ptr == nullptr)
    {
        mappedBuffer_ptr.reset(new MessageMappedBuffer());
    }

    sbeUtils.PrintMsg(mappedBuffer_ptr->getMappedBuffer(), MessageMappedBuffer::bufferLength);

    return orderBookDispatcher.dispatch(
        (char*)mappedBuffer_ptr->getMappedBuffer(), 
        0, 
        MessageMappedBuffer::bufferLength
    );
}

void MainLoopAgent::onStart() 
{
    logger->info("onStart");
}

int MainLoopAgent::doWork() 
{
    return signalSubsciption.poll();
}

void MainLoopAgent::onClose() 
{
    logger->info("onClose");
}

std::string MainLoopAgent::roleName() 
{
    return "MainLoopAgent";
}